pkg/querier/remote_read.go

```diff
@@ -78,50 +79,67 @@ func remoteReadSamples(
 	q storage.Queryable,
 	w http.ResponseWriter,
 	req *prompb.ReadRequest,
+	maxConcurrency int,
 	logger log.Logger,
 ) {
 	resp := prompb.ReadResponse{
 		Results: make([]*prompb.QueryResult, len(req.Queries)),
 	}
-	// Fetch samples for all queries in parallel.
-	errCh := make(chan error)
 
-	for i, qr := range req.Queries {
-		go func(i int, qr *prompb.Query) {
-			start, end, minT, maxT, matchers, hints, err := queryFromRemoteReadQuery(qr)
-			if err != nil {
-				errCh <- err
-				return
-			}
+	type queryJob struct {
+		index int
+		query *prompb.Query
+	}
 
-			querier, err := q.Querier(int64(start), int64(end))
-			if err != nil {
-				errCh <- err
-				return
+	jobs := make([]queryJob, len(req.Queries))
+	for i, qr := range req.Queries {
+		jobs[i] = queryJob{
+			index: i,
+			query: qr,
+		}
+	}
+	closers := make([]io.Closer, len(jobs))
+	defer func() {
+		for _, closer := range closers {
+			if closer != nil {
+				_ = closer.Close()
 			}
+		}
+	}()
 
-			seriesSet := querier.Select(ctx, false, hints, matchers...)
-
-			// We can over-read when querying, but we don't need to return samples
-			// outside the queried range, so can filter them out.
-			resp.Results[i], err = seriesSetToQueryResult(seriesSet, int64(minT), int64(maxT))
-			errCh <- err
-		}(i, qr)
+	concurrencyLimit := maxConcurrency
+	if concurrencyLimit <= 0 {
+		concurrencyLimit = len(jobs)
```

Comment:
[nit] Technically not needed due to https://github.com/grafana/dskit/blob/main/concurrency/runner.go#L101 but I suppose it doesn't hurt